### ISSUE TITLE
[camera] recording video of 200MS duration is not possible on android

### ISSUE THREAD

**Issue by wdcs-dimilkalathiya** [flutter/flutter#154167]
### Steps to reproduce

Test 1:
1. Start recording video using ```await controller?.startVideoRecording()``` 
2. Wait till 200 MS
3. Call ```controller?.stopVideoRecording()```

Test 2:
Test 1:
1. Start recording video using ```await controller?.startVideoRecording()``` 
2. Wait till 2 seconds
3. Call ```controller?.stopVideoRecording()```

### Expected results

Test 1:
- Video with 200MS runtime should be generated

Test 2:
- Video with 2 seconds runtime should be generated

### Actual results

Test 1:
- .temp file is returned with 0 byte size

Test 2:
- .temp file is returned (if extension change to .mp4 it will be video with 1.7 seconds)

### Code sample

<details open><summary>Code sample</summary>
Test 1:
```dart
    await controller?.startVideoRecording();
    await Future.delayed(const Duration(milliseconds: 200), () {});
    final video = await controller!.stopVideoRecording();
    video.path.logI;
    (await video.length()).logI;
```

Test 2:
```dart
    await controller?.startVideoRecording();
    await Future.delayed(const Duration(milliseconds: 2000), () {});
    final video = await controller!.stopVideoRecording();
    video.path.logI;
    (await video.length()).logI;
```
</details>

### Logs

<details open><summary>Logs</summary>
Test 1:
```console
I/flutter (25587): │ 💡 /data/user/0/com.zizle.dev/cache/REC7265390614749397427.temp
I/flutter (25587): │ 💡 0
```

Test 2:
```console
I/flutter (25587): │ 💡 /data/user/0/com.zizle.dev/cache/REC1589270164385580655.temp
I/flutter (25587): │ 💡 4187259
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.22.3, on macOS 14.4.1 23E224 darwin-x64, locale
    en-US)
    • Flutter version 3.22.3 on channel stable at /Users/tagline/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b0850beeb2 (6 weeks ago), 2024-07-16 21:43:41 -0700
    • Engine revision 235db911ba
    • Dart version 3.4.4
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/tagline/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.10+0-17.0.10b1087.21-11609105)

[✓] VS Code (version 1.92.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.94.0

[✓] Connected device (4 available)
    • IV2201 (mobile)         • PRHYZHS8WC6LUOTO          • android-arm64  •
      Android 13 (API 33)
    • iPhone 12 Mini (mobile) • 00008101-001039AC0AF1003A • ios            • iOS
      15.4.1 19E258
    • macOS (desktop)         • macos                     • darwin-x64     •
      macOS 14.4.1 23E224 darwin-x64
    • Chrome (web)            • chrome                    • web-javascript •
      Google Chrome 128.0.6613.84

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by wdcs-dimilkalathiya** [comment]
My observation:
- It seems that ```await controller?.startVideoRecording();``` 's future ends before actual recording starts so small videos does not get recorded and videos with longer duration comes out with less runtime than expected (eg 2 second recording time & 1.7 second output)

**Comment by huycozy** [comment]
This looks similar to https://github.com/flutter/flutter/issues/143049. Are you using [camera_android_camerax](https://pub.dev/packages/camera_android_camerax)? If not, can you try using it to see if it works well?

**Comment by wdcs-dimilkalathiya** [comment]
Hii @huycozy, thanks for the quick response

And yes i have tried with both with and without [camera_android_camerax](https://pub.dev/packages/camera_android_camerax)

camerax implementation only seems to suppress error and does not resolve issue

with camerax:
- file with 0B size will be returned if you record small videos (200-250 MS or below)
- generates smaller video than actual record time  (eg: 1.7 second runtime if you record exactly 2 seconds)

without camerax:
- throws below mentioned error if you record small videos (200-250 MS or below)
```console
CameraException(error, stop failed.)
```
- generates smaller video than actual record time  (eg: 1.7 second runtime if you record exactly 2 seconds)

https://github.com/flutter/flutter/issues/143049 suggests this issue is resolved but it only seems to be suppressed by passing empty file (0B size)

furthermore above mentioned issue is also reproducible with example app in which video_player will throw error due to invalid file (due to 0B size)

**Comment by huycozy** [comment]
I checked this with `camera_android_camerax: ^0.6.8+3` and also saw the issue. Reproduced on Flutter 3.24.1.

I think this is similar to https://github.com/flutter/flutter/issues/143049. Perhaps the camera is delayed when start recording which causes the output video to be cut off on the beginning part.

Here are the outputs with 200ms and 2000ms duration tests:

```console
// tried a few times, same result

I/flutter (17173): length 0
```

```console
// 1st time
I/flutter (17173): length 1404814

// 2nd time
I/flutter (17173): length 1434026

// 3rd time
I/flutter (17173): length 1429736
```

**Comment by reidbaker** [comment]
Tagging @camsim99 (who is on vacation this week) to correct me but with camera_android_camerax ran in to this in the initial authoring of the plugin. The issue is actually one of underlying apis. We are delaying returning the future until after the appropriate lifecycle events from camerax. (at least the best I can tell) But the callbacks do not align to actual frames being recorded.

If you or someone else can spot an error in our implementation we are happy to accept a pull request. 
See packages/camera/camera_android_camerax/example/integration_test/integration_test.dart in https://github.com/flutter/packages/pull/6697/files#diff-e23a8a3e8bd4f6038bdc550f758743f66b8e14bc9c6ae75b9432dee9bf530f28 from where we have an integration test that covers start, pause, stop.

**Comment by camsim99** [comment]
> The issue is actually one of underlying apis. We are delaying returning the future until after the appropriate lifecycle events from camerax. (at least the best I can tell) But the callbacks do not align to actual frames being recorded.

Agreed here. The only thing I could think of that explains this issue is if the Android API does not actually start recording when the methods to start recording complete. We await for an event from Android to say that video recording has started, then allow the Flutter method to start video capturing to return (see [source](https://github.com/flutter/packages/blob/71e827e7df6833e6942873361f189adcb041d7f1/packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart#L1094-L1096)).

Definitely open to a fix though if someone spots something off. As an aside, I do wonder if using CameraX without Flutter versus say Camera2 would allow this. That may be good grounds to follow up in a bug against CameraX.

**Comment by camsim99** [comment]
Closing this as a duplicate of https://github.com/flutter/flutter/issues/148138 for clarity, but feel free to re-open if you feel that this issue is separate.
